mailpoet/mailpoet
# Automation filters [MAILPOET-4946] [MAILPOET-4419] [MAILPOET-5002]

## Description

This PR:
1. Finalizes subject `fields`.
2. Adds support for field `filters` on triggers (definition, saving, validation, UI, filtering logic).
3. Implements two filters (`string` and `enum_array`).
4. Adds `Subscriber segments` filter.
5. Adds test coverage.
6. Adds a few small fixes for issues encountered on the way.

Note that filters will work in the free plugin itself, but it's only premium that allows adding, removing, or modifying them. This is consistent with how we treat automation steps.

## Code review notes

I opted for adding a new `filters` step field. Does that make sense? At the moment, we use the filters only for triggers. I'm guessing we could reuse the same `filters` field as well for conditional steps, for instance.

## QA notes

Please test with and without the premium plugin and make sure that the subscriber segment filter works with all condition types. Please also check that updating a site with already existing and active automations works.

**Please, ping @JanJakes before merging.**

## Linked PRs

https://github.com/mailpoet/mailpoet-premium/pull/704

## Linked tickets

[MAILPOET-4946] (Filters, Subscriber segments filter)
[MAILPOET-4419] (String filter)
[MAILPOET-5002] (Enum array filter)

## After-merge notes

_N/A_


[MAILPOET-4946]: https://mailpoet.atlassian.net/browse/MAILPOET-4946?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[MAILPOET-4419]: https://mailpoet.atlassian.net/browse/MAILPOET-4419
[MAILPOET-5002]: https://mailpoet.atlassian.net/browse/MAILPOET-5002
